### PR TITLE
Implement commands to upgrade/refund large amount of stats

### PR DESCRIPTION
**Pull request by KurzedMetal**
Hey guys,

Players at high levels have been struggling on changing builds with the current commands that increase stats by 1. When they have 100+ stats it gets a bit annoying and probably stressful for the server, so I implemented some functions to improve this.

The old ones still work: `/up_flow` `/up_strength`
"upgrade" is accepted too: `/upgrade_flow` `/upgrade_strength`
I added a "refund" command: `/refund_flow` `/refund_strength`
And now amount can be added to any of them too: `/up_flow_10` `/refund_flow_25`

You may ask, "why isn't the parameter separated by space?", well it seems that with the current API of Telegram isn't possible to create a touchable/clickable link with a parameter separated by space, so using just underscore to separate the parameter let you link the parameter properly. IMO that could be useful at some point.

I'll appreciate any feedback for possible improvements.

GL guys.

### CODE REVIEW

File: src/game/handlers/improveStats.js [hunk 1]
```diff
@@ -7,32 +10,26 @@ import { ObjectId } from 'mongodb'
...
+import { statUpgradeCost, unspentStatPoints, statIds } from './statHelpers'
+import {
+  showBonus, equippedIds, equipDetails, mergeEquipBonuses,
+} from './equipHelpers'
 
-function buttons (char, statNames) {
-  return statNames.map(statName =>
-    `:heavy_plus_sign: /up_${statName} (-${getStatCost(char, statName)})`
+function statUpgradeButtons (char) {
+  return toPairs(statIds).map(({ 0: statName, 1: statId }) =>
```
Comment (derekstavis) [inline src/game/handlers/improveStats.js]: This is hacky. There's a proper syntax for destructuring arrays:

```
return toPairs(statIds).map(([ statName, statId ]) =>
```

File: src/game/handlers/improveStats.js [hunk 2]
```diff
@@ -7,32 +10,26 @@ import { ObjectId } from 'mongodb'
...
-      'constitution',
-      'reflex',
-      'accuracy',
-      'flow',
-    ]).concat([
-      ':arrow_left: /overworld',
-    ]),
+  return splitEvery(
+    2,
+    statUpgradeButtons(char).concat([':arrow_left: /overworld']),
```
Comment (derekstavis) [inline src/game/handlers/improveStats.js]:
```js
[ ...statUpgradeButtons(char), ':arrow_left: /overworld', ]
```

File: src/game/handlers/improveStats.js [hunk 3]
```diff
@@ -7,32 +10,26 @@ import { ObjectId } from 'mongodb'
 
 import membersExp from '../core/membersExp'
 import { level } from '../core/level'
-import { getStatCost, getCurrentStatPoints } from './statHelpers'
+import { statUpgradeCost, unspentStatPoints, statIds } from './statHelpers'
+import {
+  showBonus, equippedIds, equipDetails, mergeEquipBonuses,
```
Comment (derekstavis) [inline src/game/handlers/improveStats.js]: Split in multiple lines

File: src/game/handlers/overworld.js [hunk 4]
```diff
@@ -5,8 +5,10 @@ import {
 function view (_) {
   return [
     _(':globe_with_meridians: Welcome to the overworld! :globe_with_meridians:'),
-    _('\nOur telegram group is @telemmo, join and chat with us!'),
-    _('To know more about this bot, send /about'),
+    '',
+    _('Our telegram group is @telemmo, join and chat with us!'),
+    '',
+    _('/about to know more about this bot'),
```
Comment (derekstavis) [inline src/game/handlers/overworld.js]: Could be in a separate PR, for traceability.

File: src/game/handlers/upStat.js [hunk 5]
```diff
@@ -11,19 +12,22 @@ import { level } from '../core/level'
...
-    $set: { [statId]: amount + 1 },
+    $set: { [statId]: currentStatScore + changeAmount },
   })
 }
 
 export default function call (dao, provider, _, msg) {
-  const statName = msg.matches[1]
+  const statName = msg.matches[2]
   const statId = statIds[statName]
+  const changeType = msg.matches[1] === 'refund' ? 'refund' : 'upgrade'
```
Comment (derekstavis) [inline src/game/handlers/upStat.js]: I'd prefer validating (error message) over assuming  'upgrade' for anything different than `refund`.

Comment (KurzedMetal) [inline src/game/handlers/upStat.js]: The only way to call that handler right now is "refund", "up" and "upgrade" because of the regex in `route.js`. That line you reviewed is mostly to streamline "up" as "upgrade" so I don't have to be checking "up or upgrade" everywhere else.

I guess this could be better for handling future modifications that what it is in the current PR

    const changeType = msg.matches[1] === 'up' ? 'upgrade' : msg.matches[1]

do you like that a bit more? or did you have anything else in mind?

File: src/game/handlers/improveStats.js (hunk 1 above)
Comment (KurzedMetal) [inline src/game/handlers/improveStats.js]: Didn't know about this :P
I'm kinda new to JS specially ES6.

File: src/game/handlers/improveStats.js (hunk 2 above)
Comment (KurzedMetal) [inline src/game/handlers/improveStats.js]: Looks definitely better.

File: src/game/handlers/improveStats.js (hunk 1 above)
Comment (derekstavis) [inline src/game/handlers/improveStats.js]: No problem! You are performing very well!

**Comment by KurzedMetal** [comment]
@derekstavis 
I just did some history editing 😈  and split the PR like you asked and applied all recommendations :)

File: src/game/handlers/overworld.js (hunk 4 above)
Comment (KurzedMetal) [inline src/game/handlers/overworld.js]: I just created #107 for most of the output cleanup changes.

**Comment by KurzedMetal** [comment]
Just an extra detail I had in my mind but didn't have time to implement.

Now the /upgrade_x_x commands give info when it fails because the player lacks points.

telegram_2017-03-17_16-29-48

**Comment by MarcoWorms** [comment]
Hey @KurzedMetal, i want to thank you a lot for your PRs again. I've merged the ones you did this week but this specific one is giving me a conflict that i think you can solve it better, could you rebase this for te current origin/master? Thanks!! Love ya <3

**Comment by KurzedMetal** [comment]
You should've merged this one (the more complex) first.

I'm out tonight. When I get back I'll check it

**Comment by MarcoWorms** [comment]
Thanks! Sorry about that :(

**Comment by KurzedMetal** [comment]
No prob dude. Fixed it already :)

**Comment by MarcoWorms** [comment]
Thanks! I'll test this locally then merge, we will deploy your changes this week (we just need to merge with the current deployed branch that runs the game in more than 1 core)

**Comment by KurzedMetal** [comment]
I tested everything already, but feel free to double check, it's always a
good thing
